k4Reco/GaudiLumiCalClusterer/include/GlobalMethodsClass.h

```diff
@@ -0,0 +1,139 @@
+    ZLayerThickness,
+    ZLayerPhiOffset,
+    ZLayerZOffset,
+    ThetaMin,
+    ThetaMax,
+    LogWeightConstant,
+    MoliereRadius,
+    MinSeparationDist,
+    ElementsPercentInShowerPeakLayer,
+    NumOfNearNeighbor,
+    ClusterMinNumHits,
+    MinHitEnergy,
+    MinClusterEngyGeV,
+    MiddleEnergyHitBoundFrac,
+    WeightingMethod,
+    Signal_to_GeV,
+    BeamCrossingAngle,
+    LumiInColName,
+    BetaGamma,
+    Gamma
+  };
+  //
+
+  enum Coordinate_t { COTheta, COPhi, COZ, COR, COP, COA };
+
+  static std::string GetParameterName(Parameter_t par);
+
+  typedef std::map<Parameter_t, int> ParametersInt;
+  typedef std::map<Parameter_t, double> ParametersDouble;
+  typedef std::map<Parameter_t, std::string> ParametersString;
+
+  GlobalMethodsClass();
+  GlobalMethodsClass(const GlobalMethodsClass& rhs) = delete;
+  GlobalMethodsClass& operator=(const GlobalMethodsClass& rhs) = default;
+  GlobalMethodsClass(GlobalMethodsClass&& rhs) = default;
+  GlobalMethodsClass& operator=(GlobalMethodsClass&& rhs) = default;
+  ~GlobalMethodsClass() = default;
+
+  void SetConstants(const std::map<std::string, std::variant<int, float, std::string>>& _lcalRecoPars);
+  WeightingMethod_t getMethod(const std::string& methodName) const;
+
+  WeightingMethod_t method = LogMethod;
+
+  double _backwardRotationPhi;
+  ParametersInt GlobalParamI;
+  ParametersDouble GlobalParamD;
+  ParametersString GlobalParamS;
+
+  double toSignal(const double valNow) const;
+  double toGev(const double valNow) const;
+
+  void ThetaPhiCell(const int cellId, std::map<GlobalMethodsClass::Coordinate_t, double>& thetaPhiCell) const;
+
+  static void CellIdZPR(const int cellId, int& cellZ, int& cellPhi, int& cellR, int& arm);
+  static int CellIdZPR(const int cellZ, const int cellPhi, const int cellR, const int arm);
+  static int CellIdZPR(const int cellId, const Coordinate_t ZPR);
+
+  void PrintAllParameters() const;
+  void initializeAdditionalParameters();
+
+  static double posWeight(const double cellEngy, const double totEngy,
+                          const GlobalMethodsClass::WeightingMethod_t method, const double logWeightConstNow);
+
+  inline double getCalibrationFactor() const { return GlobalParamD.at(Signal_to_GeV); }
+
+  template <class T, class U>
+  inline void rotateToGlobal(const T* loc, U* glob) const;
+  void rotateToLumiCal(const edm4hep::Vector3f& glob, double* loc) const;
+
+  std::tuple<std::optional<edm4hep::MutableCluster>, std::optional<edm4hep::MutableReconstructedParticle>>
+  getLCIOObjects(const LCCluster& thisClusterInfo, const double minClusterEnergy, const bool cutOnFiducialVolume,
+                 const edm4hep::CalorimeterHitCollection& calohits) const;
+
+private:
+  bool SetGeometryDD4HEP();
+
+  // LumiCal rotations angles ( local->Global )
+  std::map<int, double> m_armCosAngle{};
+  std::map<int, double> m_armSinAngle{};
+};
+
+template <class T, class U>
+void GlobalMethodsClass::rotateToGlobal(const T* loc, U* glob) const {
```

Comment:
```suggestion
std::array<U, 3> GlobalMethodsClass::rotateToGlobal(const std::array<T, 3>& loc) const {
```
I would prefer interfaces like this, i.e. without in-out parameters. I don't know how much work it is though.